libraries/core_libs/consensus/src/dag/dag.cpp

```diff
@@ -584,25 +596,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
   assert(dag_order_set.count(new_anchor));
   addToDag(new_anchor, blk_hash_t(), vec_blk_t(), 0, true);
 
+  std::pair<std::unordered_map<blk_hash_t, std::shared_ptr<DagBlock>>, std::vector<trx_hash_t>>
+      expired_dag_blocks_to_remove;
   for (auto &v : non_finalized_blocks) {
-    for (auto &blk : v.second) {
-      if (dag_order_set.count(blk) == 0) {
-        auto dag_block = dag_blk_mgr_->getDagBlock(blk);
+    for (auto &blk_hash : v.second) {
+      if (dag_order_set.count(blk_hash) == 0) {
+        auto dag_block = dag_blk_mgr_->getDagBlock(blk_hash);
         auto pivot_hash = dag_block->getPivot();
-        addToDag(blk, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+
+        if (validateBlockNotExpired(dag_block, expired_dag_blocks_to_remove)) {
+          addToDag(blk_hash, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+        }
```

Comment:
this could be refactored to:

if (validateBlockNotExpired(dag_block, ct1)) {
  addToDag(blk_hash, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
} else {
 //  add dag_block->getTrxs() to ct2
}